File: spaced_rep_march_2025/inorderTraversal.swift

public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init() { self.val = 0; self.left = nil; self.right = nil; }
    public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
    public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
        self.val = val
        self.left = left
        self.right = right
    }
}
 
class Solution {
    func inorderTraversal(_ root: TreeNode?) -> [Int] {
        guard var n = root else {
            return []
        }
        var stack: [Int] = []
        var order: [Int] = []
        var curr: TreeNode? = n

        while stack.count > 0 || curr != nil {
            while curr != nil && var currNode = curr.left {
                stack.push(curr)
                curr = currNode
            }
            if curr == nil {
                curr = stack.removeLast()
            }
            order.append(curr.val)
            curr = curr.right
        }

        return order
    }
}

File: spaced_rep_july_2023/inorderTraversal.js
var inorderTraversal = function(root) {
    if (!root) return []
    return [...inorderTraversal(root.left), root.val, ...inorderTraversal(root.right)]    
};

var inorderTraversal = function(root) {
    let stack = [], curr = root, inorder = []
    while (curr || stack.length) {
        while (curr && curr.left) {
            stack.push(curr)
            curr = curr.left
        }
        if (!curr) curr = stack.pop()
        inorder.push(curr.val)
        curr = curr.right
    }
    return inorder
};